system-helper/flatpak-system-helper.c

```diff
@@ -141,6 +371,343 @@ no_progress_cb (OstreeAsyncProgress *progress, gpointer user_data)
+        }
+    }
+
+  return FALSE;
+}
+
+static gboolean
+handle_get_child_repo_for_pull (FlatpakSystemHelper   *object,
+                                GDBusMethodInvocation *invocation,
+                                const gchar           *ref,
+                                const gchar           *commit,
+                                const gchar           *arg_installation)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(FlatpakDir) system = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *bindfs_path = NULL;
+
+
+  g_debug ("GetChildRepoForPull %s", arg_installation);
+
+  system = dir_get_system (arg_installation, &error);
+  if (system == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  bindfs_path = g_find_program_in_path ("bindfs");
+  if (bindfs_path == NULL)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED,
+                                             "Can't find bindfs in the system.");
+      return TRUE;
+    }
+  else
+    {
+      g_autoptr(FlatpakSystemHelperOngoingPull) new_pull = NULL;
+      g_autofree gchar *child_repo_path = NULL;
+      g_autofree gchar *user_dir = NULL;
+      g_autofree gchar *root_dir = NULL;
+      g_autofree gchar *flatpak_dir = NULL;
+      g_autofree gchar *arg_user = NULL;
+      g_autofree gchar *out_location = NULL;
+      guint uid_pull_count = 0;
+      uid_t uid;
+
+      if (!get_connection_uid (invocation, &uid, &error))
+        {
+          g_dbus_method_invocation_return_gerror (invocation, error);
+          return TRUE;
+        }
+
+      uid_pull_count = get_uid_pull_count (uid);
+      if (uid_pull_count > MAX_PULLS_PER_USER)
+        {
+          g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_LIMITS_EXCEEDED,
+                                                 "Maximum number of pulls per user exceeded for UID: %u", uid);
+          return TRUE;
+        }
+
+      flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+      g_autofree gchar *repo_tmp = g_build_filename (flatpak_dir, "repo", "tmp", NULL);
+
+      g_auto(GStrv) parts = flatpak_decompose_ref (ref, NULL);
+      if (check_partial_pull_availability (repo_tmp, parts[1], commit, &out_location))
+        {
+          g_autoptr(GFile) user_dir_file = NULL;
+
+          g_debug ("Found a partial pull for ref %s at: %s", parts[1], out_location);
+          root_dir = g_path_get_dirname (out_location);
+          user_dir = g_strdup_printf ("%s-bindfs", root_dir);
+          user_dir_file = g_file_new_for_path (user_dir);
+          if (!g_file_query_exists (user_dir_file, NULL))
+            g_mkdir_with_parents (user_dir, 0755);
+          child_repo_path = g_steal_pointer (&out_location);
+        }
+      else
+        {
+          g_autoptr(OstreeRepo) child_repo = NULL;
```

Comment:
This right here, spits out a glib-critical. For some reason, this is having problems while `g_autoptr`-ized cleanup. I wonder if the problem is around `flatpak_dir_create_child_repo` below (given that I have changed function to a private API in this same commit)

bt:
```
F: GetChildRepoForPull 
F: Checking partial pull availability for ref org.gnome.Platform (commit: f6b4bc140bb2e2b8f72d1d2ae2109e4f951dfabfbe59102e22ba200ed2edeadf)

(/usr/libexec/flatpak-system-helper:2378): GLib-GObject-CRITICAL **: g_object_unref: assertion 'G_IS_OBJECT (object)' failed

Thread 5 "pool" received signal SIGTRAP, Trace/breakpoint trap.
[Switching to Thread 0x7fffe4812700 (LWP 2453)]
0x00007ffff538d4c1 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
(gdb) bt
#0  0x00007ffff538d4c1 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#1  0x00007ffff538e7bd in g_logv () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007ffff538e91f in g_log () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#3  0x00007ffff6e3ec0f in  () at /lib/x86_64-linux-gnu/libostree-1.so.1
#4  0x00007ffff5669f8a in g_object_unref ()
    at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#5  0x0000555555570938 in glib_autoptr_cleanup_OstreeRepo (_ptr=<synthetic pointer>) at /usr/include/ostree-1/ostree-autocleanups.h:55
#6  0x0000555555570938 in handle_get_child_repo_for_pull (object=0x555555808760, invocation=<optimized out>, ref=0x7fffdc001bb0 "runtime/org.gnome.Platform/x86_64/3.30", commit=0x7fffdc001b60 "f6b4bc140bb2e2b8f72d1d2ae2109e4f951dfabfbe59102e22ba200ed2edeadf", arg_installation=<optimized out>)
    at system-helper/flatpak-system-helper.c:630
#7  0x00007fffef5ce038 in ffi_call_unix64 ()
    at /lib/x86_64-linux-gnu/libffi.so.6
#8  0x00007fffef5cda9a in ffi_call () at /lib/x86_64-linux-gnu/libffi.so.6
#9  0x00007ffff566575e in g_cclosure_marshal_generic ()
    at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#10 0x00007ffff5664f25 in g_closure_invoke ()
    at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#11 0x00007ffff56776e2 in  () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#12 0x00007ffff567f311 in g_signal_emitv ()
    at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
---Type <return> to continue, or q <return> to quit---
#13 0x00005555555b6064 in _flatpak_system_helper_skeleton_handle_method_call (connection=<optimized out>, sender=<optimized out>, object_path=<optimized out>, interface_name=0x7fffd80011b0 "org.freedesktop.Flatpak.SystemHelper", method_name=0x7fffd8005d70 "GetChildRepoForPull", parameters=<optimized out>, invocation=0x55555581a910, user_data=0x555555808760) at common/flatpak-dbus-generated.c:6780
#14 0x00007ffff597cb7e in  () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#15 0x00007ffff59337fd in  () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#16 0x00007ffff53b076e in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#17 0x00007ffff53afd75 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#18 0x00007ffff5126494 in start_thread (arg=0x7fffe4812700)
    at pthread_create.c:333
#19 0x00007ffff4e68acf in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
(gdb) 
```
@wjt @pwithnall  Any idea?

Reply:
Nothing jumps out, sorry. Check out the [Debugging reference count problems](https://developer.gnome.org/gobject/stable/tools-refdb.html) tips from the GObject docs.

Reply:
Nothing jumps out at me either. Try running under `valgrind` and see if it prints any other messages about memory corruption or reading/writing unallocated heap memory. The real problem might be caused by heap corruption elsewhere, and this is only a symptom of it. Make sure to use the [`glib.supp` suppressions file from GLib](https://gitlab.gnome.org/GNOME/glib/blob/master/glib.supp) with `--suppressions=glib.supp` to ignore known false positives.

Reply:
I got it.

I didn't use the `ostree_repo_get_path` API properly. It has `[transfer none]` and I was assigning:
```
g_autoptr(GFile) repo_file = ostree_repo_get_path (child_repo);
```
Fixing it, fixes the critical too.
the `child_repo` is g_autoptr-ized too and hence `repo_file` pointer was being double-freed.

Thanks!
I will push fixup!